### PR TITLE
Refactor

### PR DESCRIPTION
**Pull request by thesabinelim**
- Firebase functions now abstracted into a new module db.py
- All lines now 79 chars or less

### CODE REVIEW

File: botcore/db.py [hunk 1]
```diff
@@ -0,0 +1,162 @@
...
+        Retrieve entries for all unverified members from the database.
+
+        Returns:
+            Dictionary where each key is a member ID and each value is the info
+            associated with that member.
+        """
+
+        unverified = {}
+        docs = self._get_members_col() \
+            .where("id_verified", "==", False).stream()
```
Comment (TerryAg) [inline botcore/db.py]: Comparing boolean values with equality? I'm not sure what the `.where()` function wants as arguments but is this best practice?

File: botcore/db.py [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+            id: ID of a secret.
+
+        Returns:
+            Secret bytes associated with id or None if no such secret exists.
+        """
+
+        doc = self._get_secrets_col().document(str(id)).get().to_dict()
+        if doc is None:
+            return None
+        return doc["secret"]
```
Comment (TerryAg) [inline botcore/db.py]: A lot of your code assumes the schema in firebase is intact. What happens if someone breaks it via the UI? Not something you need to consider now, just a thought

File: botcore/perms.py [hunk 3]
```diff
@@ -1,83 +1,62 @@
...
-            if error:
-                raise NotVerified("You must be verified to do that.")
-            return False
-        return True
-    return check(predicate)
+        member = _get_member(cog, ctx.author.id)
+        if member is None or config["verified-role"] \
+            not in _get_role_ids(member):
+            return False, "You must be verified to do that."
+        return True, None
```
Comment (TerryAg) [inline botcore/perms.py]: not a fan of returning a tuple for the sake of being able to return multiple values. consider a custom class. (PCSoc discord bot handles this well)

File: botcore/perms.py [hunk 4]
```diff
@@ -169,100 +137,75 @@ def is_admin_user(error=False):
...
     def predicate(cog, ctx):
-        member = cog.bot.get_guild(config["server-id"]).get_member(ctx.author.id)
-        if set(config["admin-roles"]).isdisjoint(map(lambda r: r.id, member.roles)):
-            if error:
-                raise NotAdminUser("You are not authorised to do that.")
-            return False
-        return True
-    return check(predicate)
+        member = _get_member(cog, ctx.author.id)
+        if set(config["admin-roles"]).isdisjoint(_get_role_ids(member)):
```
Comment (TerryAg) [inline botcore/perms.py]: What is happening here? This is something that looks like could be extracted into a helper function that has a much clearer function name "is_user_admin(member)" or something like that

File: botcore/db.py (hunk 1 above)
Comment (thesabinelim) [reply botcore/db.py]: This is how the Firebase SDK works, which I don't really like either :(

File: botcore/db.py (hunk 2 above)
Comment (thesabinelim) [reply botcore/db.py]: This is a good point. I'll make this an issue

File: botcore/perms.py [hunk 5]
```diff
@@ -1,83 +1,62 @@
...
-    pass
-
-class NotAdminChannel(CheckFailure):
-    pass
-
-class NotDMChannel(CheckFailure):
-    pass
-
-def check(predicate):
+def check(predicate, error):
```
Comment (VineBalloon) [inline botcore/perms.py:6]: Doesn't the discord python library already provide a [check decorator](https://discordpy.readthedocs.io/en/latest/ext/commands/api.html#ext-commands-api-checks)?

File: botcore/verify.py [hunk 6]
```diff
@@ -1,84 +1,98 @@
...
+        self._secret = None
         self.state_handler = [
-            self.verify_state_await_name,
-            self.verify_state_await_unsw,
-            self.verify_state_await_zid,
-            self.verify_state_await_email,
-            self.verify_state_await_code,
-            self.verify_state_await_id,
-            self.verify_state_await_approval
+            self.state_await_name,
```
Comment (VineBalloon) [inline botcore/verify.py:38]: Try to avoid unnecessary name changes, it clutters the PR

File: botcore/verify.py [hunk 7]
```diff
@@ -95,170 +109,201 @@ def fetch_verifying(self):
...
     def get_code(self, user):
         msg = bytes(str(user.id), "utf8")
         return hmac.new(self.secret, msg, "sha256").hexdigest()
 
-    async def verify_proc_begin(self, member):
+    async def proc_begin(self, member):
         if member.id in self.verifying:
-            await member.send("You are already undergoing the verification process. To restart, type `!restart`.")
+            await member.send("You are already undergoing the verification "
+                "process. To restart, type `!restart`.")
```
Comment (VineBalloon) [inline botcore/verify.py:122]: Hardcoding the command prefix even though it exists in your config? I would just remove the config option, we're not changing the prefix anytime soon.

**Pull request merged**